pkg/operator/resource/resourceapply/monitoring_test.go

```diff
@@ -99,49 +153,80 @@ func TestApplyServiceMonitor(t *testing.T) {
-	_, modified, err := ApplyServiceMonitor(context.TODO(), dynamicClient, events.NewInMemoryRecorder("monitor-test"), required)
-	if err != nil {
-		t.Fatalf("unexpected error: %v", err)
-	}
-	if !modified {
-		t.Fatalf("expected the service monitor will be modified, it was not")
-	}
-
-	if len(dynamicClient.Actions()) != 2 {
-		t.Fatalf("expected 2 actions, got %d", len(dynamicClient.Actions()))
-	}
-
-	updateAction, isUpdate := dynamicClient.Actions()[1].(clienttesting.UpdateAction)
-	if !isUpdate {
-		t.Fatalf("expected second action to be update, got %+v", dynamicClient.Actions()[1])
-	}
-	updatedMonitorObj := updateAction.GetObject().(*unstructured.Unstructured)
-
-	labels, _, err := unstructured.NestedStringMap(updatedMonitorObj.UnstructuredContent(), "spec", "selector", "matchLabels")
-	if err != nil {
-		t.Fatalf("unable to get selector: %v", err)
-	}
-
-	expectedKeys := []string{"app", "custom"}
-	resultKeys := []string{}
-	for key := range labels {
-		resultKeys = append(resultKeys, key)
-	}
-	sort.Strings(resultKeys)
-
-	if !reflect.DeepEqual(resultKeys, expectedKeys) {
-		t.Fatalf("expected %#v selectors, got %#v", expectedKeys, resultKeys)
+	for _, tc := range []struct {
+		name          string
+		existing      string
+		expectModify  bool
+		expectActions sets.Set[string]
+	}{
+		{
+			name:          "same label, same spec",
+			existing:      serviceMonitor,
+			expectActions: sets.New("get"),
+		},
+		{
+			name:          "different label, same spec",
+			existing:      serviceMonitorDifferentLabelSameSpec,
+			expectModify:  true,
+			expectActions: sets.New("get", "update"),
+		},
+		{
+			name:          "same label, different spec",
+			existing:      serviceMonitorSameLabelDifferentSpec,
+			expectModify:  true,
+			expectActions: sets.New("get", "update"),
+		},
+		{
+			name:          "different label, different spec",
+			existing:      serviceMonitorDifferentLabelDifferentSpec,
+			expectModify:  true,
+			expectActions: sets.New("get", "update"),
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			dynamicClient := dynamicfake.NewSimpleDynamicClient(dynamicScheme, readServiceMonitorFromBytes([]byte(tc.existing)))
+			required := resourceread.ReadUnstructuredOrDie([]byte(serviceMonitor))
+			_, modified, err := ApplyServiceMonitor(context.TODO(), dynamicClient, events.NewInMemoryRecorder("monitor-test"), required)
+			if err != nil {
+				t.Fatalf("unexpected error: %v", err)
+			}
+			if !modified && tc.expectModify {
+				t.Fatalf("expected the service monitor to be modified, it was not")
+			}
+
+			actions := dynamicClient.Actions()
+			if len(actions) != len(tc.expectActions) {
+				t.Fatalf("expected %d actions, got %d: %v", tc.expectActions.Len(), len(actions), actions)
+			}
+			actionSet := sets.New[string]()
+			for _, action := range actions {
+				actionSet.Insert(action.GetVerb())
+			}
+			if !tc.expectActions.Equal(actionSet) {
+				t.Fatalf("expected actions %v, got %v", tc.expectActions, actionSet)
+			}
```

Comment:
Casting actions' verbs to a set here to account for flakes such as the one below (owing to order guarantees).
```

```